File: app/src/main/java/com/example/ogan/codedenim/login/LoginActivity.java
package com.example.ogan.codedenim.login;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.annotation.TargetApi;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.text.TextUtils;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.AutoCompleteTextView;
import android.widget.Button;
import android.widget.EditText;

import com.example.ogan.codedenim.MainActivity;
import com.example.ogan.codedenim.NetworkConnectivity;
import com.example.ogan.codedenim.R;
import com.example.ogan.codedenim.ServiceGenerator;
import com.example.ogan.codedenim.UserCategory;
import com.example.ogan.codedenim.sessionManagement.UserSessionManager;

import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;


/**
 * A login screen that offers login via email/password.
 */
public class LoginActivity extends AppCompatActivity {

    // User Session Manager Class
    UserSessionManager session;

    // UI references.
    private AutoCompleteTextView mEmailView;
    private EditText mPasswordView;
    private View mProgressView;
    private View mLoginFormView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        // User Session Manager
        session = new UserSessionManager(getApplicationContext());

        // Set up the login form.
        mEmailView = findViewById(R.id.email);
        mPasswordView = findViewById(R.id.password);


        Button mEmailSignInButton = findViewById(R.id.email_sign_in_button);
        mEmailSignInButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if(NetworkConnectivity.INSTANCE.checkNetworkConnecttion(getApplicationContext())){
                    attemptLogin();
                } else {
                    NetworkConnectivity.INSTANCE.showNoInternetMessage(LoginActivity.this);
                }
            }
        });

        Button mRegisterButton = findViewById(R.id.email_register_button);
        mRegisterButton.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(getApplicationContext(), UserCategory.class);
                startActivity(intent);
            }
        });

        mLoginFormView = findViewById(R.id.login_form);
        mProgressView = findViewById(R.id.login_progress);
    }


    /**
     * Attempts to sign in or register the account specified by the login form.
     * If there are form errors (invalid email, missing fields, etc.), the
     * errors are presented and no actual login attempt is made.
     */
    private void attemptLogin() {

        // Reset errors.
        mEmailView.setError(null);
        mPasswordView.setError(null);

        // Store values at the time of the login attempt.
        final String email = mEmailView.getText().toString();
        final String password = mPasswordView.getText().toString();

        boolean cancel = false;
        View focusView = null;

        // Check for a valid password, if the user entered one.
        if (TextUtils.isEmpty(password)) {
            mPasswordView.setError(getString(R.string.error_field_required));
            focusView = mPasswordView;
            cancel = true;
        }

        // Check for a valid email address.
        if (TextUtils.isEmpty(email)) {
            mEmailView.setError(getString(R.string.error_field_required));
            focusView = mEmailView;
            cancel = true;
        } else if (!isEmailValid(email)) {
            mEmailView.setError(getString(R.string.error_invalid_email));
            focusView = mEmailView;
            cancel = true;
        }

        if (cancel) {
            // There was an error; don't attempt login and focus the first
            // form field with an error.
            focusView.requestFocus();
        } else {
            // Show a progress spinner, and kick off a background task to
            // perform the user login attempt.
            showProgress(true);

            Call<ResponseBody> login = ServiceGenerator.INSTANCE.getApiMethods().loginCorper(email, password, "password");
            login.enqueue(new Callback<ResponseBody>() {
                @Override
                public void onResponse(@NonNull Call<ResponseBody> call, @Nullable Response<ResponseBody> response) {

                    showProgress(false);

                    if (response != null) {
                        if (response.isSuccessful()) {
                            session.createUserLoginSession(email, password);
                            finish();
                            Intent intent = new Intent(getApplicationContext(), MainActivity.class);
                            startActivity(intent);
                        }

                        System.out.println(response.message());
                        System.out.println(response.code());
                    }


                }

                @Override
                public void onFailure(@NonNull Call<ResponseBody> call, @Nullable Throwable t) {
                    showProgress(false);
                }
            });
        }
    }

    private boolean isEmailValid(String email) {
        return email.contains("@");
    }

    private boolean isPasswordValid(String password) {
        return password.length() > 4;
    }

    /**
     * Shows the progress UI and hides the login form.
     */
    @TargetApi(Build.VERSION_CODES.HONEYCOMB_MR2)
    private void showProgress(final boolean show) {
        // On Honeycomb MR2 we have the ViewPropertyAnimator APIs, which allow
        // for very easy animations. If available, use these APIs to fade-in
        // the progress spinner.
        int shortAnimTime = getResources().getInteger(android.R.integer.config_shortAnimTime);

        mLoginFormView.setVisibility(show ? View.GONE : View.VISIBLE);
        mLoginFormView.animate().setDuration(shortAnimTime).alpha(
                show ? 0 : 1).setListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationEnd(Animator animation) {
                mLoginFormView.setVisibility(show ? View.GONE : View.VISIBLE);
            }
        });

        mProgressView.setVisibility(show ? View.VISIBLE : View.GONE);
        mProgressView.animate().setDuration(shortAnimTime).alpha(
                show ? 1 : 0).setListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationEnd(Animator animation) {
                mProgressView.setVisibility(show ? View.VISIBLE : View.GONE);
            }
        });
    }
}

File: app/src/main/java/com/example/ogan/codedenim/login/LoginActivity.kt
package com.example.ogan.codedenim.login

import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.text.TextUtils
import android.util.Patterns
import android.view.View
import android.widget.*
import com.example.ogan.codedenim.*
import com.example.ogan.codedenim.sessionManagement.UserSessionManager
import okhttp3.ResponseBody
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response




/**
 * A login screen that offers login via email/password.
 */
class LoginActivity : AppCompatActivity() {

    // User Session Manager Class
    private lateinit var session: UserSessionManager

    // UI references.
    private lateinit var mEmailView: AutoCompleteTextView
    private lateinit var mPasswordView: EditText
    private lateinit var mProgressView: View
    private lateinit var mLoginFormView: View
    private lateinit var mEmailSignInButton: Button

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)

        // User Session Manager
        session = UserSessionManager(applicationContext)

        //title = "Sign In"

        // Set up the login form.
        mEmailView = findViewById(R.id.email)
        mPasswordView = findViewById(R.id.password)


        mEmailSignInButton = findViewById(R.id.email_sign_in_button)
        mEmailSignInButton.setOnClickListener {
            if (MyUtilClass.checkNetworkConnection(applicationContext)) {
                attemptLogin()
            } else {
                MyUtilClass.showNoInternetMessage(this@LoginActivity)
            }
        }

        val mRegisterTextView = findViewById<TextView>(R.id.register_tv)
        mRegisterTextView.setOnClickListener {
            val intent = Intent(applicationContext, UserCategory::class.java)
            startActivity(intent)
        }

        mLoginFormView = findViewById(R.id.login_form)
        mProgressView = findViewById(R.id.login_progress)
    }


    /**
     * Attempts to sign in or register the account specified by the login form.
     * If there are form errors (invalid email, missing fields, etc.), the
     * errors are presented and no actual login attempt is made.
     */
    private fun attemptLogin() {

        // Reset errors.
        mEmailView.error = null
        mPasswordView.error = null

        // Store values at the time of the login attempt.
        val email = mEmailView.text.toString()
        val password = mPasswordView.text.toString()

        var cancel = false
        var focusView: View? = null

        // Check for a valid email address.
        if (TextUtils.isEmpty(email)) {
            mEmailView.error = getString(R.string.error_field_required)
            focusView = mEmailView
            cancel = true
        } else if (!Patterns.EMAIL_ADDRESS.matcher(email).matches()) {
            mEmailView.error = getString(R.string.error_invalid_email)
            focusView = mEmailView
            cancel = true
        }

        // Check for a valid password, if the user entered one.
        if (TextUtils.isEmpty(password)) {
            mPasswordView.error = getString(R.string.error_field_required)
            focusView = mPasswordView
            cancel = true
        }

        if (cancel) {
            // There was an error; don't attempt login and focus the first
            // form field with an error.
            focusView?.requestFocus()
        } else {
            // Show a progress spinner, and kick off a background task to
            // perform the user login attempt.
            showProgress()
            login(email, password)
        }
    }

    private fun login(email: String, password: String) {
        ServiceGenerator.apiMethods.loginCorper(email, password, "password")
                .enqueue(object : Callback<ResponseBody> {
                    override fun onResponse(call: Call<ResponseBody>, response: Response<ResponseBody>?) {

                        hideProgress()

                        if (response != null) {
                            if (response.isSuccessful) {
                                session.createUserLoginSession(email, password)
                                finish()
                                val intent = Intent(applicationContext, MainActivity::class.java)
                                startActivity(intent)
                            } else
                                MyUtilClass.showErrorMessage(this@LoginActivity, response)

                            println(response.message())
                            println(response.code())
                        }


                    }

                    override fun onFailure(call: Call<ResponseBody>, t: Throwable?) {
                        hideProgress()
                        if (t != null)
                            Toast.makeText(applicationContext, t.localizedMessage, Toast.LENGTH_LONG)
                                    .show()

                    }
                })
    }

    private fun showProgress(){
        mProgressView.visibility = View.VISIBLE
        mEmailSignInButton.isEnabled = false
    }

    private fun hideProgress(){
        mProgressView.visibility = View.GONE
        mEmailSignInButton.isEnabled = true
    }
}

